scenarios.md

```diff
@@ -31,6 +31,53 @@ Notes:
 | `https://e.com`<br>w/o OI  | `https://a.e.com`<br>w/o OI | `https://b.e.com`<br>w/ OI | Main and A in `Site{https://e.com}`<br>B in `Origin{https://b.e.com}` |
 
+
+## Session history scenarios
+
+The previous scenarios illustrate some cases where the proposed algorithm ignores the origin isolation request because a pre-existing site-keyed agent cluster exists for a same-origin URL in the current frame tree. The following scenario illustrates a case where the session history contributes to this calculation.
+
+### Navigating a subframe
+
+* Main frame: `https://e.com` w/ OI
+* Subframe: `https://x.e.com` w/o OI
+* Outcome (per above): Main in `Origin{https://e.com}` / Sub in `Site{https://e.com}`
+
+The user clicks a link in the `https://x.e.com` subframe which takes them to `https://e.org`.
+
+The user then clicks a button in the main frame which inserts a second subframe, pointing at `https://x.e.com`. But in the meantime the server operator has deployed origin isolation on `https://x.e.com`. Thus the scenario is now:
+
+* Main frame: `https://e.com` w/ OI
+* Subframe 1: `https://e.org` (with `https://x.e.com` in the session history)
+* Subframe 2: `https://x.e.com` w/ OI
+
+In this case, subframe 2 ends up keyed by `Site{https://e.com}`. This ensures that if the user navigates subframe 1 back, restoring the `Site{https://e.com}`-keyed instance of `https://x.e.com` in subframe 1, that subframe 1 and subframe 2 are still in the same `Site{https://e.com}` agent cluster, i.e. we have avoided isolating same-origin pages from each other.
+
+### Inserting iframes and saving JS references
+
+* Main frame: `https://e.com` w/ OI
+* Subframe: `https://e.org` w/o OI
+* Outcome (per above): Main in `Origin{https://e.com}` / Sub in `Site{https://e.org}`.
+
+JavaScript code in the main frame goes through a variety of contortions:
+
+```js
+// Save a reference to the sub-frame window.
+window.savedFrame = frames[0];
+
+// Now remove it from the DOM.
+frames[0].frameElement.remove();
+```
+
+Some time later, JavaScript code inserts a new subframe, again pointing at `http://e.org`. But in the meantime the server operator has deployed origin isolation on `https://e.org`. Thus the scenario is now:
+
+* Main frame: `https://e.com` w/ OI
+* Subframe: `https://e.org` w/ OI
+
+Will the subframe get site-keyed, or origin-keyed?
+
+The answer is origin-keyed. The `window.savedFrame` variable does mean that the agent cluster map still contains an entry with key `Site{https://e.org}`, which itself contains the saved realm and corresponding `Window` object. However, because the iframe was removed from the DOM, the `Window` has no browsing context, and thus the browsing context group does not contain any browsing contexts that have seen the `https://e.org` origin. Thus the request for origin isolation is respected, and we end up with `Origin{https://example.org/}` as the key.
```

Comment:
But if the parent keeps the agent alive wouldn't that keep the agent cluster alive and so if you go find your agent cluster why would it create a new one?

Reply:
Yeah, I think you're right... Although I'm not sure if we're using the same reasoning.

I now think basically each BCG should have an append-only list of historical agent cluster keys. So, removing a BC should not change this list.

I'll reformulate the PR in these terms...

Reply:
I think that's what I proposed originally and you all didn't like that 😊

Reply:
Oh, sorry I wasn't clear; I meant the last few commits to align on your proposed model, since you kept poking holes in ours :)